### PR TITLE
Implement external_command_helper and add real_rc_string back

### PR DESCRIPTION
**Pull request by mattdurak**

### CODE REVIEW

File: tests/external_command_helper_ut/external_command_helper_ut.c [hunk 1]
```diff
@@ -51,6 +74,105 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
...
+{
+    ASSERT_IS_TRUE(pclose_is_pending, "mocked version of pclose should not have been called without a matching popen");
+
+    // Close and delete the temp file
+    int result = fclose(stream);
+    remove(temp_file_name);
+
+    ASSERT_ARE_EQUAL(int, 0, result);
+
+    LogInfo("Deleted temp file: %s", temp_file_name);
```
Comment (dcristoloveanu) [inline tests/external_command_helper_ut/external_command_helper_ut.c:102]: >LogInfo [](start = 4, length = 7)

Requires the xlogging include

File: tests/external_command_helper_ut/external_command_helper_ut.c [hunk 2]
```diff
@@ -51,6 +74,105 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
...
+static FILE* last_opened_file_handle;
+static bool pclose_is_pending;
+
+static int pclose_override_return;
+static int hook_pclose(FILE* stream)
+{
+    ASSERT_IS_TRUE(pclose_is_pending, "mocked version of pclose should not have been called without a matching popen");
+
+    // Close and delete the temp file
+    int result = fclose(stream);
```
Comment (dcristoloveanu) [inline tests/external_command_helper_ut/external_command_helper_ut.c:97]: >fclose [](start = 17, length = 6)

I'd assert directly on this, no result variable needed

File: tests/external_command_helper_ut/external_command_helper_ut.c [hunk 3]
```diff
@@ -51,6 +74,105 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
...
+    pclose_is_pending = true;
+
+    // Pick a file name
+    ASSERT_IS_NOT_NULL(tmpnam(temp_file_name));
+
+    // Open file to fill with test data
+    FILE* temp = fopen(temp_file_name, "w");
+
+    int fputs_result = fputs(test_data_to_report_as_output, temp);
+    ASSERT_IS_TRUE(fputs_result >= 0);
```
Comment (dcristoloveanu) [inline tests/external_command_helper_ut/external_command_helper_ut.c:131]: >ASSERT_IS_TRUE( [](start = 4, length = 15)

Can assert without an intermediate variable.

File: tests/external_command_helper_ut/external_command_helper_ut.c [hunk 4]
```diff
@@ -51,6 +74,105 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
...
+
+    int fclose_result = fclose(temp);
+    ASSERT_ARE_EQUAL(int, 0, fclose_result);
```
Comment (dcristoloveanu) [inline tests/external_command_helper_ut/external_command_helper_ut.c:134]: Same here

File: tests/external_command_helper_ut/external_command_helper_ut.c [hunk 5]
```diff
@@ -51,6 +74,105 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
...
+
+    // Open file for caller
+    result = fopen(temp_file_name, mode);
+    last_opened_file_handle = result;
```
Comment (dcristoloveanu) [inline tests/external_command_helper_ut/external_command_helper_ut.c:138]: Same here

**Review by dcristoloveanu — APPROVED** [review]
:shipit:

File: tests/external_command_helper_ut/external_command_helper_ut.c (hunk 5 above)
Comment (mattdurak) [reply tests/external_command_helper_ut/external_command_helper_ut.c:138]: This is saving the handle. And now I see it's not being used like it was in the other tests from block_storage_2 (but it should be used). Fixing

---
In reply to: [554284116](https://github.com/Azure/c-util/pull/93#discussion_r554284116) [](ancestors = 554284116)

**Pull request merged**